feat(mobile): collectible details content

pete-watters opened the pull request:
This PR updates collectible details pages to show more relevant information and manipulate the data to make it look better in the UI

✨ Features:

- [x] Adds a new feature flag for collectible details specifically
- [x] Adds code to show relevant metadata in a nicely formulated way
- [x] Fixes some problems raised by @alexp3y in a previous PR

☎️ Demo:

https://github.com/user-attachments/assets/fc450a4d-6434-4f1d-ac3c-119997608bc3

📔 Improvements:

I made some judgements on what to show and where to show it. This is open for interpretation and can be changed or improved but for now it's functional

pete-watters committed:
chore(mobile): gate collectible details seperately

pete-watters committed:
chore(mobile): use promise.all to group calls

pete-watters committed:
chore(mobile): clean up mapping of API data

pete-watters committed:
feat(mobile): populate collectible details with relevant information

pete-watters committed:
fix(mobile): fix bug rendering bns names

pete-watters committed:
feat(mobile): filter out floor price

pete-watters commented on apps/mobile/src/features/token/utils/render-collectible-details.tsx:
```diff
@@ -0,0 +1,120 @@
+  // If value is a URL, make it clickable
+  if (typeof stringValue === 'string' && stringValue.startsWith('https://')) {
+    return (
+      <Pressable onPress={() => Linking.openURL(stringValue)}>
```
I am using `Linking` for this as when I tried to use `useOpenURL` I was getting a warning as this isn't a hook/ function component.  I can refactor it if needed but I think it's OK for now as it works the same. 

```
 React Hook "useOpenURL" is called in function "renderValue" that is neither a React function component nor a custom React Hook function
```

pete-watters commented on packages/services/src/infrastructure/api/gamma/gamma-api.utils.ts:
```diff
@@ -0,0 +1,71 @@
+import { z } from 'zod';
```
@alexp3y I updated this as per recommendation [here](https://github.com/leather-io/mono/pull/1635#discussion_r2362743029). Let me know if you recommend more improvements

alexp3y replied:
Looks good but this probably isn't the best spot for these map utils. 

API clients shouldn't really be concerned with anything outside of network interactions and their own API types.

The service layer acts as the adapter that takes the output from this (and any other) client calls and creates the `Sip9Asset`. So the asset services (and their utils) should be responsible for knowing how to map from the provider API types to our internal domain model.

pete-watters commented on apps/mobile/src/features/token/utils/render-collectible-details.tsx:
```diff
@@ -0,0 +1,120 @@
+    // Skip undefined or empty string values
+    if (value === undefined || value === '') return;
+
+    const keysToSkip = ['description', 'floorPrice', 'floor_price_amount'];
```
@alexp3y I left out `floor_price` for now. We were discussing it [here](https://github.com/leather-io/mono/pull/1635#discussion_r2362716204). If we really want it I can convert it to `Money` but for now I've just filtered it out

pete-watters replied:
I decided to just exclude this from our model altogether for now until we decide we want to show it

alexp3y commented on packages/models/src/assets/asset.model.ts:
```diff
@@ -209,6 +209,59 @@ export type SupportedSip9ContentType =
     >
   | '';
+
+export interface Sip9Collection {
```
All of these `Sip9Asset` subtypes (`Sip9Collection`, `Sip9Owner`, etc...) could be moved out to a separate `sip9-asset.model.ts` file to reduce noise here.

alexp3y commented on packages/models/src/assets/asset.model.ts:
```diff
@@ -209,6 +209,59 @@ export type SupportedSip9ContentType =
+  id: string;
+  type?: string;
+  name: string;
+  isVerified: boolean;
```
Is verified by whom?

pete-watters replied:
I have no clue, by Gamma I assume 🤷

pete-watters replied:
Ditching this

alexp3y commented on packages/models/src/assets/asset.model.ts:
```diff
@@ -209,6 +209,59 @@ export type SupportedSip9ContentType =
+  type?: string;
+  name: string;
+  isVerified: boolean;
+  locationUrl: string;
```
Location of what?

pete-watters replied:
It's the collection main page location on `Gamma`

pete-watters replied:
I've changed this to be `collectionExplorerUrl` and made it work as a clickable link now. That was a bug you helped me fix

https://github.com/user-attachments/assets/d1371651-ce79-45c2-a35a-55ee7036259e

alexp3y commented on packages/models/src/assets/asset.model.ts:
```diff
@@ -209,6 +209,59 @@ export type SupportedSip9ContentType =
+  };
+}
+
+export interface Sip9Owner {
```
This seems more like Gamma profile data than purely representing a SIP-9 owner?

pete-watters replied:
I am ditching this and not showing it

alexp3y commented on packages/models/src/assets/asset.model.ts:
```diff
@@ -209,6 +209,59 @@ export type SupportedSip9ContentType =
+
+export interface Sip9Attribute {
+  traitType: string;
+  displayType?: string;
```
What is a `traitType` or `displayType`?

pete-watters replied:
These map to `attribute_groups.attributes.type` and `attribute_groups.attributes.label` in the Gamma API. 

Different NFTs can have individual traits for example a monkey with a hat. 

In the Hiro API these fields are called `trait_type` and `display_type` so I used those names.  

- `displayType` / `label` is the type of attribute the NFT has
- `traitType` is a string usually on Gamma and is often set as `simple`. 

So for example for this [Mutant Monkey](https://gamma.io/stacks/nfts/SP125J1ADVYWGWB9NQRCVGKYAG73R17ZNMV17XEJ7.mutant-monkeys_2396) 

This is what we get for attribute groups:

```
[
    {
        "title": "Attributes",
        "attributes": [
            {
                "type": "simple",
                "label": "Background",
                "value": "MM1 Yellow",
                "rarity_percent_of_100": 7.65
            },
            {
                "type": "simple",
                "label": "Clothes",
                "value": "MM1 Green Tshirt",
                "rarity_percent_of_100": 4.61
            },
            {
                "type": "simple",
                "label": "Eyes",
                "value": "MM1 Eye Patch",
                "rarity_percent_of_100": 6.66
            },
            {
                "type": "simple",
                "label": "Fur",
                "value": "MM1 Pink",
                "rarity_percent_of_100": 6.08
            },
            {
                "type": "simple",
                "label": "Hat",
                "value": "MM1 Viking Helmet",
                "rarity_percent_of_100": 3.6
            },
            {
                "type": "simple",
                "label": "Level",
                "value": "1",
                "rarity_percent_of_100": 45.42
            },
            {
                "type": "simple",
                "label": "Mouth",
                "value": "MM1 Pizza",
                "rarity_percent_of_100": 7.29
            },
            {
                "type": "simple",
                "label": "SGR",
                "value": "MM1 100%",
                "rarity_percent_of_100": 67.19
            },
            {
                "type": "simple",
                "label": "Trait Count",
                "value": "6",
                "rarity_percent_of_100": 40.55
            }
        ]
    }
]
```

`displayType` seems to have more potential values but I didn't handle them exhaustively:
```
displayType

Defines how the trait value should be rendered or interpreted by marketplaces and wallets. Common display types include:

string (default) - Regular text value
number - Numeric value
boost_number - Numeric boost/modifier (often shown with +/-)
boost_percentage - Percentage boost (shown with %)
date - Unix timestamp to be displayed as a date
```

pete-watters replied:
I've renamed these as:

- attributeTraitType
- attributeValue
- attributeRarityPercent

Hopefully that gives better semantic meaning

alexp3y replied:
Nice, thanks for explaining. Didn't realize this was actually part of the SIP-016 standard.

alexp3y commented on packages/models/src/assets/asset.model.ts:
```diff
@@ -209,6 +209,59 @@ export type SupportedSip9ContentType =
+  traitType: string;
+  displayType?: string;
+  value: any;
+  rarityPercent?: number;
```
Rarity according to what/whom?

pete-watters replied:
It's the collection attribute group rarity percentage according to Gamma

pete-watters replied:
So to elaborate using [this example](https://github.com/leather-io/mono/pull/1655#discussion_r2413157714)

That monkey is one of `4.61%` in the collection with a Green t-shirt 

<img width="812" height="261" alt="Screenshot 2025-10-08 at 13 22 33" src="https://github.com/user-attachments/assets/471f09f8-a89d-45b9-ac1f-e54d39c46243" />

pete-watters replied:
I fixed a bug where these weren't actually rendering properly. 

Now the NFT attributes list properly if available

https://github.com/user-attachments/assets/454bb898-ef9b-481f-b677-a4a1ecf05da7

They look kind of ugly though. I will get a design review and try get some help rendering them in a nicer way using labels or something

kyranjamie commented on apps/mobile/src/features/token/utils/render-collectible-details.tsx:
```diff
@@ -0,0 +1,120 @@
+    return t`Collection URL`;
+  }
+
+  // Replace underscores with spaces
+  let formatted = key.replace(/_/g, ' ');
+
+  // Add spaces between PascalCase words
+  formatted = formatted.replace(/([a-z])([A-Z])/g, '$1 $2');
```
why not chain? wrapped in fns for readability


```ts
function replaceUnderscores(str: string) {
  return str.replace(/_/g, ' ');
}

function addPascalCaseSpaces(str: string) {
  return str.replace(/([a-z])([A-Z])/g, '$1 $2');
}

const formatted = addPascalCaseSpaces(replaceUnderscores("My_TestKey"));
```

pete-watters replied:
I'll change this to make it easier to follow. I let Claude create this whole section. 

These days I am veering less towards human readability and increased code generation

kyranjamie commented on apps/mobile/src/features/token/utils/render-collectible-details.tsx:
```diff
@@ -0,0 +1,120 @@
+  return formatted.charAt(0).toUpperCase() + formatted.slice(1);
+}
+
+function renderValue(value: any, key: string): React.ReactNode {
```
```suggestion
function renderValue(value: unknown, key: string): React.ReactNode {
```

better to use unknown if narrowing

kyranjamie commented on apps/mobile/src/features/token/utils/render-collectible-details.tsx:
```diff
@@ -0,0 +1,120 @@
+
+  // Defined sort order for the collectible details
+  const sortOrder = ['name', 'tokenId', 'assetId', 'contractId', 'mimetype', 'contentType'];
+  getSortedEntries(obj, sortOrder).forEach(([key, value]) => {
```
Could use a map?

kyranjamie commented on apps/mobile/src/features/token/utils/render-collectible-details.tsx:
```diff
@@ -0,0 +1,120 @@
+
+    const label = formatLabel(key);
+
+    if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
```
```suggestion
    if (isObject(value) && value !== null && !Array.isArray(value)) {
```

kyranjamie commented on packages/services/src/infrastructure/api/gamma/gamma-api.utils.ts:
```diff
@@ -0,0 +1,71 @@
+export function mapGammaCollection(
+  collection?: GammaNftMetadata['item']['collection']
+): Sip9Collection | undefined {
+  if (!collection) return undefined;
```
is it legit that collection can be `undefined`, or is this just to handle async data that's missing?

IMO map is a slightly confusing term as it could indicate an iteration.

If absolutely necessary, wrapping, vibe coded example

```ts
export const optionalize =
  <A, R>(fn: (a: A) => R) =>
  (a: A | undefined): R | undefined =>
    a === undefined ? undefined : fn(a);

// Example
export function mapGammaCollection(
  collection: GammaNftMetadata['item']['collection']
): Sip9Collection {
  return {
    id: collection.id,
    type: collection.type,
    name: collection.name,
    isVerified: collection.is_verified,
    locationUrl: collection.location_url,
    totalItems: collection.total_items,
    floorPrice: collection.floor_price_amount
      ? { amount: collection.floor_price_amount.amount, unit: collection.floor_price_amount.unit }
      : undefined,
  };
}

export const mapGammaCollectionSafe = optionalize(mapGammaCollection);
```

pete-watters replied:
The response data is sometimes inconsistent, so there could be an `undefined` collection. I can update it as per recommendation, thanks!

alexp3y commented on packages/models/src/assets/asset.model.ts:
```diff
@@ -209,6 +209,59 @@ export type SupportedSip9ContentType =
       | 'model/gltf-binary'
     >
   | '';
+
```
Didn't catch the addition of the content type model above, but the `Sip9ContentType` seems very exhaustive - Where exactly does this list come from? What's the purpose of the `SupportedSip9ContentType`? Doesn't look like we reference it or do any validation checks anywhere.

pete-watters replied:
I added that list in [this PR](https://github.com/leather-io/mono/pull/1642/files#r2382514401). It is very exhaustive, I added it as a reference to all possible types but maybe I should remove it? I got claude to tell me all the possible permutations a SIP9 could be. 

`SupportedSip9ContentType` is the ones I was able to actually test myself and make sure they work OK. The thinking was that if someone has a collectible that's not showing then it would be easy to extend `SupportedSip9ContentType` by adding the new type. 

I use `SupportedSip9ContentType` as a type in `sip9.native.tsx` 
```
interface MediaInfo {
  contentType: SupportedSip9ContentType;
  isVideo: boolean;
  isImage: boolean;
  isAudio: boolean;
}
```

In there it's used to help decide which component to render based on `ContentType`. 

I'm happy to go with your recommendation on what's best here

alexp3y replied:
Makes sense. I think the thing that feels off for me is that in the mapper we cast any value we get to `SupportedSip9ContentType`, whether it's supported or not.

Maybe whether something is "supported" is actually a concern of the specific UI that's displaying it? So, the model `contentType` field just holds any possible `Sip9ContentType` and then the supported determination is made elsewhere.

pete-watters committed:
chore(mobile): address PR feedback

alexp3y commented on packages/models/src/assets/asset.model.ts:
```diff
@@ -209,6 +209,59 @@ export type SupportedSip9ContentType =
+  rarityPercent?: number;
+}
+
+export interface Sip9Details {
```
Solid foundation here on the `Sip9Asset` model, but it still feels pretty coupled to Gamma's API (e.g., `rarityRank`, `floorPrice`, `isVerified` still directly mirror their structure).

We should try to delineate between the on-chain fields that represent the essence of a SIP-9 (e.g., `name`, `contractId`, `tokenId`, `imageUri`) and any enriched data computed by a specific platform/marketplace (e.g., `rarityRank`, `floorPrice`, `isVerified`).

One idea could be to prefix enriched fields with something like "marketplace" or "provider" (e.g., `marketplaceRarityRank`, `marketplaceFloorPrice`), or just encapsulate them all in a parent structure like `Sip9ProviderData`. 

Even then, we should approach the "marketplace"/"provider" data like a universal remote - We should design it to work for Gamma now but also, in theory, so that it can plug into any provider in the future without needing changes.

pete-watters replied:
My thinking here was to get something working quickly that we can refine later. I'll see what I can do to improve it

alexp3y reviewed (commented):
Great start pulling together the additional SIP-9 data. Added some comments that build on the idea of making the `Sip9Asset` more provider-agnostic.

Also think we could refine which fields get presented in the UI before releasing, feels like a bit of a data dump atm. Fields we want to present and naming conventions there could also be a good guiding force for the model and help reduce any Gamma specificity.

pete-watters commented:
> Great start pulling together the additional SIP-9 data. Added some comments that build on the idea of making the `Sip9Asset` more provider-agnostic.
> 
> Also think we could refine which fields get presented in the UI before releasing, feels like a bit of a data dump atm. Fields we want to present and naming conventions there could also be a good guiding force for the model and help reduce any Gamma specificity.

Thanks @alexp3y . I'll see what I can do to clean it up. 

My idea here was that this is good enough for now and we could have it added behind a feature flag. Then we can test it out and see what fields we actually want to show and what to exclude. I already filtered out a lot of them based on a short conversation with @camerow but had planned to discuss it more with the design team. 

It is 100% a data dump for now as I believe that's good enough for a first iteration until we decide how best to refine it or if we even want to dedicate more time to that at all

pete-watters committed:
chore(mobile): move metadata manipulation to asset service

pete-watters committed:
chore(mobile): clean up SIP9 model and only model data we use

pete-watters committed:
fix(mobile): add handling of nested arrays of objects to auto print nft attributes

pete-watters committed:
fix(mobile): encapsulate SIP9 provider data to keep the interface cleaner

pete-watters committed:
fix(mobile): fix duplicate key error message

pete-watters commented:
@alexp3y & @kyranjamie - thanks for the feedback and patience here. 

I have made changes as requested and I think it's looking a lot better now so I appreciate the tips.  If you spot anything else you think I can improve on let me know!

@alexp3y : 
- I moved the data manipulation to the `asset` service and renamed them from `map` to `transform`
- I got rid of a lot of fields we probably don't need
- Others I re-labelled so they are more intuitive
- I wrapped the additional fields in their own interface - `Sip9ProviderData`

alexp3y commented on packages/models/src/assets/sip9-asset.model.ts:
```diff
@@ -0,0 +1,128 @@
+  details: Sip9Details;
+}
+
+export interface Sip9Asset extends BaseNonFungibleCryptoAsset {
```
Awesome improvement. Looks a lot cleaner. Seeing this now I'm kind of second-guessing the Provider-speicifc structure and wondering if a lot of this could be consolidated further (as there's a lot of repitition). Happy to keep as-is for now but lmk if you have time for a quick chat later and we can discuss.

pete-watters commented:
Thanks @alexp3y  and @kyranjamie 👍 

@alexp3y and I had a pair session today and discussed further improvements to the models. I am working on that now but am going to merge this PR first as it's already pretty big. I'll add a fast follow up with model changes.
